ansible/tasks/nginx/main.yml

```diff
@@ -4,6 +4,85 @@
     name: 
       - nginx
 
+- name: Setup stream in nginx.conf
+  register: nginxconf
+  copy:
+    dest: /etc/nginx/nginx.conf
+    content: |
+      user www-data;
+      worker_processes auto;
+      pid /run/nginx.pid;
+      include /etc/nginx/modules-enabled/*.conf;
+
+      events {
+              worker_connections 768;
+              # multi_accept on;
+      }
+      
+      http {
+              ##
+              # Basic Settings
+              ##
+
+              sendfile on;
+              tcp_nopush on;
+              tcp_nodelay on;
+              keepalive_timeout 65;
+              types_hash_max_size 2048;
+              # server_tokens off;
+
+              # server_names_hash_bucket_size 64;
+              # server_name_in_redirect off;
+
+              include /etc/nginx/mime.types;
+              default_type application/octet-stream;
+
+              ##
+              # SSL Settings
+              ##
+
+              ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
+              ssl_prefer_server_ciphers on;
+
+              ##
+              # Logging Settings
+              ##
+
+              access_log /var/log/nginx/access.log;
+              error_log /var/log/nginx/error.log;
+
+              ##
+              # Gzip Settings
+              ##
+
+              gzip on;
+
+              # gzip_vary on;
+              # gzip_proxied any;
+              # gzip_comp_level 6;
+              # gzip_buffers 16 8k;
+              # gzip_http_version 1.1;
+              # gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;
+
+              ##
+              # Virtual Host Configs
+              ##
+
+              include /etc/nginx/conf.d/*.conf;
+              include /etc/nginx/sites-enabled/*;
+       }
+       
+       # Forward :443 to /var/run/nginx/$domain.https.sock
+       stream {    
+              server {
+                  listen      443;
+                  proxy_protocol on;
+                  proxy_pass  unix:/var/run/nginx/$ssl_preread_server_name.https.sock;
```

Comment:
What is preventing a user from making this socket without "ownership" of the domain? Why should we use this instead of a socket in $HOME? Why would we use /var/run instead of /run?

Reply:
/run works, I didn't know /run is a common place to hold those files.

Reply:
Well technology, he cannot.
You can only listening a socket by creating it, meaning if the socket file presents, you cannot listen it without removing it. So users won't be able to hijack a socket in most cases.

By putting every domain under same folder, we don't need to regenerate and reload ngnix. This also solves the issue that multiple users trying to bind to same domain name.

Reply:
The weakness seems to be that they can claim _any_ unclaimed hostname, so they could grab something like admin.hashbang.sh if it wasn't bound already. They could create listeners for whole blocks of hostnames and eat up the namespace.

Reply:
Yes, but I guess you shouldn't use shell server for that purpose if you want a really stable bind. This is the same limitation that current ad-hoc port based http server has. Anyone can grab a port and take it forever.
So I don't think it's a big disadvantage compared to any other approaches, unless we introduce a humans to approve each domain before take the bind.

Reply:
Someone more clever than me might find a way to bind it to their username?

Reply:
ping?